frame/fast-unstake/src/lib.rs

```diff
@@ -0,0 +1,558 @@
+
+			// the range that we're allowed to check in this round.
+			let current_era = pallet_staking::CurrentEra::<T>::get().unwrap_or_default();
+			let eras_to_check = {
+				let bonding_duration = <T as pallet_staking::Config>::BondingDuration::get();
+
+				// get the last available `bonding_duration` eras up to current era in reverse
+				// order.
+				let total_check_range = (current_era.saturating_sub(bonding_duration)..=
+					current_era)
+					.rev()
+					.collect::<Vec<_>>();
+				debug_assert!(
+					total_check_range.len() <= (bonding_duration + 1) as usize,
+					"{:?}",
+					total_check_range
+				);
+
+				// remove eras that have already been checked, take a maximum of
+				// final_eras_to_check.
+				total_check_range
+					.into_iter()
+					.filter(|e| !checked.contains(e))
+					.take(final_eras_to_check as usize)
+					.collect::<Vec<_>>()
+			};
+
+			log!(debug, "{} eras to check: {:?}", eras_to_check.len(), eras_to_check);
+
+			if eras_to_check.is_empty() {
+				// `stash` is not exposed in any era now -- we can let go of them now.
+				let num_slashing_spans = Staking::<T>::slashing_spans(&stash).iter().count() as u32;
+
+				let ctrl = match pallet_staking::Bonded::<T>::get(&stash) {
+					Some(ctrl) => ctrl,
+					None => {
+						Self::deposit_event(Event::<T>::Errored { stash });
+						return <T as Config>::WeightInfo::on_idle_unstake()
+					},
+				};
+
+				let ledger = match pallet_staking::Ledger::<T>::get(ctrl) {
+					Some(ledger) => ledger,
+					None => {
+						Self::deposit_event(Event::<T>::Errored { stash });
+						return <T as Config>::WeightInfo::on_idle_unstake()
+					},
+				};
+
+				let unstake_result = pallet_staking::Pallet::<T>::force_unstake(
+					RawOrigin::Root.into(),
+					stash.clone(),
+					num_slashing_spans,
+				);
+
+				let pool_stake_result = if let Some(pool_id) = maybe_pool_id {
+					pallet_nomination_pools::Pallet::<T>::join(
+						RawOrigin::Signed(stash.clone()).into(),
+						ledger.total,
+						pool_id,
+					)
+				} else {
+					Ok(())
+				};
+
+				let result = unstake_result.and(pool_stake_result);
+				log!(
+					info,
+					"unstaked {:?}, maybe_pool {:?}, outcome: {:?}",
+					stash,
+					maybe_pool_id,
+					result
+				);
+				Self::deposit_event(Event::<T>::Unstaked { stash, maybe_pool_id, result });
+
+				<T as Config>::WeightInfo::on_idle_unstake()
+			} else {
+				// eras remaining to be checked.
+				let mut eras_checked = 0u32;
+				let is_exposed = eras_to_check.iter().any(|e| {
+					eras_checked.saturating_inc();
+					Self::is_exposed_in_era(&stash, e)
+				});
+
+				log!(
+					debug,
+					"checked {:?} eras, exposed? {}, (v: {:?}, u: {:?})",
+					eras_checked,
+					is_exposed,
+					validator_count,
+					eras_to_check.len()
+				);
+
+				// NOTE: you can be extremely unlucky and get slashed here: You are not exposed in
+				// the last 28 eras, have registered yourself to be unstaked, midway being checked,
+				// you are exposed.
```

Comment:
I thought this was impossible since the staker's been chilled right on `register_fast_unstake`?

Reply:
Very very unlikely, but not impossible. You might be exposed in era n if you are part of the snapshot that's taken a few hundred blocks before the beginning of era n (and there's one session of validators being buffered, which only complicates things further). If you register yourself for fast-unstake after this snapshot is taken, you might be exposed in the upcoming era, and thus slashed.

All in all, I delegated all of this to wallets to handle. They can and should easily check if the user for which they are trying to unbond are part of the snapshot, or if they are actually exposed or not. 

Lastly, 99% of the audience of this pallet are stakers who have never been exposed and will never be either.